### PR TITLE
Replace boost::math::special_functions

### PR DESCRIPTION
**Pull request by heisenbuug**
Implementing `digamma()` and `trigamma()` to replace `boost::math` functions.

- [x] Digamma
- [x] Trigamma
- [ ] Solve `narrow conversion` error when passing `int`
- [x] Port tests for both the functions

This gives error
`trigamma(1)`  `digamma(1)`

But this works
`trigamma(1.0)`  `digamma(1.0)`

No test cases are failing but still have to handle this.

### CODE REVIEW

File: src/mlpack/core/math/digamma.hpp
Comment (conradsnicta) [inline src/mlpack/core/math/digamma.hpp:23]: @heisenbuug change `fucntion` to `function`  
seems to be an often made spelling mistake  
https://www.dictionary.com/browse/function

File: src/mlpack/core/math/digamma.hpp
Comment (conradsnicta) [inline src/mlpack/core/math/digamma.hpp:59]: change `fucntion` to `function`

**Review by conradsnicta — CHANGES_REQUESTED** [review]

File: src/mlpack/core/math/digamma.hpp
Comment (heisenbuug) [reply src/mlpack/core/math/digamma.hpp:23]: Yup, I know the correct spelling. But somehow it feels like my fingers are programmed to make the mistake.
No excuses :sweat_smile: :sweat_smile:

File: src/mlpack/core/math/digamma.hpp
Comment (shrit) [inline src/mlpack/core/math/digamma.hpp:34]: @heisenbuug I think we should have `N==7` here, the same way you have done it to other functions right? otherwise, I imagine that this function can be executed if `N==6` and it will segfault in this case.

File: src/mlpack/core/math/digamma.hpp
Comment (shrit) [inline src/mlpack/core/math/digamma.hpp:202]:
```suggestion

```

File: src/mlpack/core/math/digamma.hpp
Comment (shrit) [inline src/mlpack/core/math/digamma.hpp:204]:
```suggestion
```

File: src/mlpack/core/math/digamma.hpp
Comment (shrit) [inline src/mlpack/core/math/digamma.hpp:206]:
```suggestion
```

File: src/mlpack/core/math/digamma.hpp
Comment (shrit) [inline src/mlpack/core/math/digamma.hpp:200]:
```suggestion
```

File: src/mlpack/core/math/digamma.hpp
Comment (shrit) [inline src/mlpack/core/math/digamma.hpp:198]:
```suggestion
```

File: src/mlpack/core/dists/gamma_distribution.cpp
Comment (shrit) [inline src/mlpack/core/dists/gamma_distribution.cpp:103]:
```suggestion
```

File: src/mlpack/core/math/trigamma.hpp
Comment (shrit) [inline src/mlpack/core/math/trigamma.hpp:183]: Can you remove empty line asterisk, the same comment as before

**Review by shrit — COMMENTED** [review]
Just minor comments, hope to see if tests are passing

File: src/mlpack/core/math/digamma.hpp
Comment (heisenbuug) [reply src/mlpack/core/math/digamma.hpp:34]: No, this function is only being used in one place.
So, this will be only called once, so no issues.
But if you you suggest that it's safe to add `N==7`, I can do that.

**Comment by heisenbuug** [comment]
> Just minor comments, hope to see if tests are passing

I don't think the failed test cases have anything to do with `digamma` or `trigamma`.

So, if not any more changes, I think we can consider this done.

File: src/mlpack/core/math/digamma.hpp
Comment (shrit) [reply src/mlpack/core/math/digamma.hpp:33]: I think it is safer to put `N==7`, otherwise it might cause errors

**Comment by heisenbuug** [comment]
@shrit added `N == 8` condition and pushed the code.

**Comment by shrit** [comment]
Thanks, I need to investigate why Windows is failing I have restarted the job and see if it passes this time, I do not see any error thought.

**Comment by heisenbuug** [comment]
> Thanks, I need to investigate why Windows is failing I have restarted the job and see if it passes this time, I do not see any error thought.

I will check where else can i remove`boost`. What do we have to do to remove this directory `mlpack/src/mlpack/core/boost_backport`.

**Comment by heisenbuug** [comment]
I think then I can remove all the files except `string_view.hpp` and `string_view_fwd.hpp`.
Also, I should just edit `CMakeLists.txt` to remove these deleted files.

**Review by shrit — APPROVED** [review]
@heisenbuug Great work, nothing to add from my side.
@zoq @rcurtin let me know what do you think.

File: src/mlpack/core/boost_backport/CMakeLists.txt
Comment (rcurtin) [inline src/mlpack/core/boost_backport/CMakeLists.txt:9]: After #3006 and this are merged, we can remove this directory entirely, I think! :tada:

File: src/mlpack/core/math/digamma.hpp
Comment (rcurtin) [inline src/mlpack/core/math/digamma.hpp:125]: I know I am being *super* picky, but do you think you can change the names to conform with the mlpack style?  (I feel a little bad asking because I know it's tedious, but I do think consistency is useful and important.  If you like, when I have time I'm happy to pass through and try to do it as suggestions.)

File: src/mlpack/core/math/trigamma.hpp
Comment (rcurtin) [inline src/mlpack/core/math/trigamma.hpp:211]:
```suggestion
#endif
```
No need for the empty line. :)

**Review by rcurtin — COMMENTED** [review]
This is awesome!  I'm surprised by how straightforward the implementation turned out to be.  I had always been intimidated by the idea of reimplementing `digamma()` and `trigamma()`.

Do you think you could also port the digamma and trigamma tests, so that we have some knowledge that our implementation is ok?  I'm not sure the tests for `GammaDistribution` will be sufficient, and it doesn't seem *too* difficult:

https://code.woboq.org/boost/boost/libs/math/test/test_digamma.hpp.html
https://code.woboq.org/boost/boost/libs/math/test/test_trigamma.cpp.html

Let me know what you think---maybe it's harder than it looks, but it doesn't look like the tests should be *that* hard?  I dunno; I didn't look too deeply. :+1:

**Review by conradsnicta — APPROVED** [review]

**Comment by zoq** [comment]
Resolved the merge conflict, let's see if the build comes back green.

**Review by shrit — CHANGES_REQUESTED** [review]
@heisenbuug I am only waiting for tests, if it is too complex, I believe we can merge this in its current state, and then open an issue if someone would like to port the tests from boost later.
Since I do not have a strong opinion on these tests @rcurtin any thoughts?

I suppose there is a bad merge happening in the CMake file that contains the old boost file. I think we need to remove this line

**Comment by heisenbuug** [comment]
> @heisenbuug I am only waiting for tests, if it is too complex, I believe we can merge this in its current state, and then open an issue if someone would like to port the tests from boost later. Since I do not have a strong opinion on these tests @rcurtin any thoughts?
> 
> I suppose there is a bad merge happening in the CMake file that contains the old boost file. I think we need to remove this line

I will port the tests by the end of tomorrow day. Is that okay?

**Comment by heisenbuug** [comment]
@shrit I think I can remove the whole `/src/mlpack/core/boost_backport` directory since there are no files in it?

**Comment by heisenbuug** [comment]
Tests are passing locally. Not sure whats happening on CI.

**Comment by shrit** [comment]
It only did not succeed in downloading armadillo,  I will start these failing tests

File: src/mlpack/core/math/trigamma.hpp
Comment (shrit) [reply src/mlpack/core/math/trigamma.hpp:211]:
```suggestion
#endif
```

**Review by shrit — APPROVED** [review]
@heisenbuug Great work, can not wait until this one gets merged, only a minor comment, otherwise nothing from my side. Thank you for another great contribution. :100:

**Comment by shrit** [comment]
@zoq @rcurtin I am merging this tomorrow, let me know if you need to give it a review, nothing from my side on this PR

**Review by zoq — APPROVED** [review]
Nice work!

**Review by shrit — APPROVED** [review]

**Pull request merged**
